### PR TITLE
[WIP] MAT: Attention Module

### PR DESCRIPTION
**Pull request by atreyamaj**
## Description
Attention module for the Molecular Attention Transformer.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by atreyamaj** [comment]
I've combined the MATAttention and MultiHeadedAttention layers by constructing the MATAttention layer within MultiHeadedAttention, as discussed earlier.

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -47,3 +49,162 @@ def __init__(self, scale, eps=1e-5):
...
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> attention = dc.models.torch_models.layers.MATAttention('softmax', 0.33, 0.33')
+  >>> self_attn_layer = dc.models.torch_models.layers.MultiHeadedAttention(dist_kernel = 'softmax', lambda_attention = 0.33, lambda_adistance = 0.33, h = 8, hsize = 1024, dropout_p = 0.1)
+  """
+
+  def __init__(self,
+               dist_kernel,
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:73]: Missing the types.

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -47,3 +49,162 @@ def __init__(self, scale, eps=1e-5):
...
+    self.lambda_distance = lambda_distance
+    self.lambda_adjacency = 1.0 - self.lambda_attention - self.lambda_distance
+    self.d_k = hsize // h
+    self.h = h
+    linear_layer = nn.Linear(hsize, hsize)
+    self.linear_layers = nn.ModuleList([linear_layer for _ in range(3)])
+    self.dropout_p = nn.Dropout(dropout_p)
+    self.output_linear = nn.Linear(hsize, hsize, output_bias)
+
+  def _singleAttention(self,
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:115]: `_single_attention` and missing types

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -47,3 +49,162 @@ def __init__(self, scale, eps=1e-5):
...
+    adj_matrix: np.ndarray
+      Adjacency matrix of the input molecule, returned from dc.feat.MATFeaturizer()
+    dist_matrix: np.ndarray
+      Distance matrix of the input molecule, returned from dc.feat.MATFeaturizer()
+    eps: float
+      Epsilon value
+    inf: float
+      Value of infinity to be used.
+    """
+
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:148]: Space not needed. If yapf complains, let it be there.

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -47,3 +49,162 @@ def __init__(self, scale, eps=1e-5):
...
+    query: torch.Tensor
+      Standard query parameter for attention.
+    key: torch.Tensor
+      Standard key parameter for attention.
+    value: torch.Tensor
+      Standard value parameter for attention.
+    mask: torch.Tensor
+      Masks out padding values so that they are not taken into account when computing the attention score.
+    """
+
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:195]: Same comment as for space above.

**Review by vsomnath — CHANGES_REQUESTED** [review]

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -14,36 +16,193 @@ class ScaleNorm(nn.Module):
...
+  References
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> attention = dc.models.torch_models.layers.MATAttention('softmax', 0.33, 0.33')
+  >>> self_attn_layer = dc.models.torch_models.layers.MultiHeadedAttention(dist_kernel = 'softmax', lambda_attention = 0.33, lambda_adistance = 0.33, h = 8, hsize = 1024, dropout_p = 0.1)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:69]: Can you provide usage examples for these layers?

File: deepchem/models/tests/test_layers.py [hunk 6]
```diff
@@ -613,6 +613,6 @@ def test_scale_norm():
   input_ar = torch.tensor([[1., 99., 10000.], [0.003, 999.37, 23.]])
```
Comment (rbharath) [inline deepchem/models/tests/test_layers.py:613]: Can you add unit tests for these new classes?

File: deepchem/models/torch_models/layers.py (hunk 5 above)
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:69]: To clarify, an example of how to invoke the layers on representative input

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -14,36 +16,193 @@ class ScaleNorm(nn.Module):
...
     self.scale = nn.Parameter(torch.tensor(math.sqrt(scale)))
     self.eps = eps
 
-  def forward(self, x):
+  def forward(self, x : torch.Tensor):
     norm = self.scale / torch.norm(x, dim=-1, keepdim=True).clamp(min=self.eps)
     return x * norm
+
+
+class MultiHeadedMATAttention(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:54]: Can you add these classes to the docs so the docstrings render?

**Review by rbharath — COMMENTED** [review]
Did my first pass!

File: deepchem/models/tests/test_layers.py (hunk 6 above)
Comment (atreyamaj) [reply deepchem/models/tests/test_layers.py:613]: Working on the tests now :+1:

File: deepchem/models/torch_models/layers.py (hunk 7 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:54]: I'll add them alongside the tests :+1:

**Review by rbharath — APPROVED** [review]
LGTM. Ready to merge in once CI looks good and once @vsomnath has a chance to take a look

**Comment by vsomnath** [comment]
I asked @atreyamaj to make one change regarding rdkit usage (added on the other PR), and then this should be good to go.

**Comment by atreyamaj** [comment]
> I asked @atreyamaj to make one change regarding rdkit usage (added on the other PR), and then this should be good to go.

I've made the change requested by @vsomnath !

**Comment by rbharath** [comment]
@atreyamaj Can you do a pass to make sure that no CI failures are related to this PR? If so, we can go ahead and merge this in

**Comment by atreyamaj** [comment]
> @atreyamaj Can you do a pass to make sure that no CI failures are related to this PR? If so, we can go ahead and merge this in

There was a Mypy error which I have fixed. I have not spotted any other errors related to this PR.

**Pull request merged**